### PR TITLE
ESP32 Bot that works in 16khz

### PR DESCRIPTION
**Pull request by zachwe**
Several changes:
- send response TTS in many small payloads rather than one small message (required due to client memory constraints)
    - modify the simple python client so it can handle this too
- Add a timeout where the server closes the websocket connection if there is a long time with no activity
- Let client specify the audio format and sample rate it wants to receive via url parameters

I have not been able to get things working smoothly in 24khz, I'm guessing due to hardware constraints.

### CODE REVIEW

**Comment by matthewkennedy5** [comment]
Can the esp32 do int32 audio? or does it need to be int16

File: openduck-py/openduck_py/routers/voice.py [hunk 1]
```diff
@@ -136,31 +146,48 @@ async def start_response(
...
-        def _inference(sentence: str):
-            audio_chunk = styletts2_inference(text=sentence)
-            
-            audio_chunk_bytes = np.int16(audio_chunk * 32767).tobytes()
+        def _inference(sentence: str, audio_format: str = "int16"):
+            audio_chunk = styletts2_inference(
+                text=sentence,
+                output_sample_rate=self.output_sample_rate,
+            )
+            if audio_format == "int32":
```
Comment (matthewkennedy5) [inline openduck-py/openduck_py/routers/voice.py:154]: Can we cast the int32 to int16 on the client before sending it to the server?

Comment (zachwe) [reply openduck-py/openduck_py/routers/voice.py]: 549cfbefe360dc9f5cd253e1c17f55a0363ef3bd

Comment (zachwe) [reply openduck-py/openduck_py/routers/voice.py]: Done here! ^

File: clients/esp32/client.ino [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+void onMessageCallback(WebsocketsMessage message) {
+    std::string rawData = message.rawData();
+
+    const size_t numSamples = rawData.size() / sizeof(int16_t);
+    std::vector<int32_t> scaledAudioData(numSamples);
+
+    // This initial scaling factor moves int16 range to int32
+    int32_t initialScalingFactor = 1 << 16;
```
Comment (matthewkennedy5) [inline clients/esp32/client.ino:109]: We can get rid of this stuff now that we're doing int16

File: clients/esp32/client.ino [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    .ws_io_num = 9, // LRCK
+    .data_out_num = 15, // Speaker output
+    .data_in_num = 17 // Microphone input
+  };
+
+  // Install and start I2S driver with your new pin configuration
+  i2s_driver_install(i2s_num, &i2s_config, 0, NULL);
+  i2s_set_pin(i2s_num, &pin_config);  
+
+  i2s_set_sample_rates(I2S_NUM_0, 22050); // Set the sample rate for your audio file.
```
Comment (matthewkennedy5) [inline clients/esp32/client.ino:186]: Why does it set the sample rate multiple times to different values? Let's have a SAMPLE_RATE const and use that everywhere

Comment (matthewkennedy5) [reply clients/esp32/client.ino:186]: Oh, I guess input audio is 16 kHz but output audio is 24 kHz

File: clients/esp32/client.ino [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+  unsigned long currentMillis = millis();
+  if (currentMillis - previousPing > pingInterval) {
+    webSocket.ping();
+    previousPing = currentMillis;
+  }
+
+  esp_err_t err = i2s_read(
+      I2S_NUM_0,
+      (void *) i2s_read_buff_int32,
```
Comment (matthewkennedy5) [inline clients/esp32/client.ino:239]: can we make this i2s_read_buff_int16() and then not need to convert it on line 248?

File: openduck-py/openduck_py/routers/voice.py [hunk 5]
```diff
@@ -278,30 +300,48 @@ async def audio_response(
     websocket: WebSocket,
     session_id: str,
     record: bool = False,
+    input_audio_format: Literal["float32", "int32", "int16"] = "float32",
+    output_sample_rate: int = 24_000,
```
Comment (matthewkennedy5) [inline openduck-py/openduck_py/routers/voice.py:304]: output_sample_rate: int = OUTPUT_SAMPLE_RATE

File: clients/esp32/client.ino (hunk 2 above)
Comment (zachwe) [reply clients/esp32/client.ino:109]: This is the code that converts from int16 to int32 (we need to use int32 audio for `i2s_write` on my hardware)

File: clients/esp32/client.ino (hunk 4 above)
Comment (zachwe) [reply clients/esp32/client.ino:239]: no - the i2s device on my hardware required int32 audio.

**Pull request merged**